In [27]:
IRdisplay::display_html('<style>.container { width:100% !important; }</style>')

<div class="alert alert-block alert-success">
<b> Jupyter para big data: (tarea 1)
</b> 

</div>

> Entrega: 15 septiembre 2019

`Camilo Gómez Narváez` (camilo.gomez@alumnos.uv.cl)

Instituto de Estadistica, Facultad de Ciencias, Universidad de Valparaíso, Chile.

<div class="alert alert-block alert-success">
<b> 
</b> 

</div>


## 1. Resumen: (sobre el conjunto de datos)

Los registros que se han seleccionado corresponden al conjunto de datos correspondientes a préstamos y prestamistas de _Kiva_ (https://www.kiva.org). Esta organización sin fines de lucro presta servicios de contacto y redes de apoyo para que personas en todo el mundo puedan financiar pequeños proyectos a personas en todo el mundo.

Específicamente, el conjunto de datos se ha obtenido desde https://www.kaggle.com/gaborfodor/additional-kiva-snapshot

<img src="Kiva-loans.jpg">

Respecto a preguntas interesantes que se pueden resolver con este conjunto de datos, se podría pensar en caracterizaciones y búsqueda de patrones de sectores en el mundo (continentes, países, ciudades, etc.) en las cuales se encuentran los donantes que más aportan o quienes regularmente lo hacen para así componer estrategias de búsqueda para nuevos prestamistas ya que la organización tiene por objetivo incentivar el préstamo de dinero a personas con escasos recursos financieros.

## 2. Importar el conjunto: (problemas asociados)

En esta sección se abarcó el problema de cargar el conjunto de datos a R debido a que el tamaño del archivo es de 2 GB.

#### 2.1 Funciones tradicionales para la lectura de archivos .CSV
Tradicionalmente, la importación de archivos de extensión .CSV se puede llevar a cabo con funciones como:

In [2]:
df1 = read.csv2("data/loans.csv",header = T,sep = ",")
head(df1)

El problema con esta función es que no está optimizada para cargar archivos con tamaños por sobre 2 GB, por lo que la alternativa que se ha encontrado para esta situación es utilizar las funciones de la librería **data.table** (https://cran.r-project.org/web/packages/data.table/index.html).

Un segundo problema que se debe considerar al cargar conjuntos de datos que sobrepasen las capacidades de R es que por ejemplo la función *read.csv2* se puede ejecutar, pero tenderá a utilizar todos los recursos del computador (RAM, SWAP, CORES) tal como se puede ver en la siguiente imágen.

<img src="01.png">

No se pudo ejecutar el comando anterior sin congelar las funciones del computador ya que la ejecución exige la utilización de la mayor cantidad de recursos que el computador posee.

#### 2.2 Funciones para cargar conjuntos de datos por sobre las capacidades de R

In [3]:
library("data.table")

In [4]:
df1 = fread("data/loans.csv",header = T,sep = ",")
head(df1)

loan_id loan_name      original_language
1 657307  Aivy           English          
2 657259  Idalia Marizza Spanish          
3 658010  Aasia          English          
4 659347  Gulmira        Russian          
5 656933  Ricky\\t       English          
6 659605  Gloria         English          
  description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
1 Aivy, 21 years of age, is single and lives in Ozamiz City, Misamis Occidental, Philippines. To help her parents meet their financial needs, she invests in a small sari-sari business (variety store). Her parents are thankful to her because of the things she has done for the family.\\r\\n\\r\\nIn the year 2013, Aivy became a member of PMPC. She believes that the institution can provide her financial assistance to access the loan amount of 5,000 PHP to buy additional inventories for her sari-sari store business. She is thankful to the cooperative in supporting her to grow in the small business she started.                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
2 Doña Idalia, esta casada, tiene 57 años de edad, vive con su esposo y sus 3 nietos en el municipio de Choloma, Cortes. Hace 7 años que se dedica a la venta de mercadería en general como ropa usada americana, lociones, calzado y otra achinería. Las ventas las realiza en forma ambulante. Ha solicitado 8,000 Lempiras para comprar ropa en general como pantalones, camisas, blusas, ropa de niño etc. para cumplir con los pedidos que tiene. Su propósito de seguir trabajando es poder ver en crecimiento su negocio.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
3 Aasia is a 45-year-old married lady and she has been running a small general store in her community for many years.  She offers a variety of grocery items at a reasonable price and has retained many customers who regularly shop at her store. Aasia makes enough to provide for her family.\\r\\n\\r\\nTo make her store fully stocked she requested a loan for 40,000 PKR which she will use to buy stock of rice, sugar and flour from the market at lower rates. In this way she would be ab

In [5]:
colnames(df1)

[1] "loan_id"                         "loan_name"                      
 [3] "original_language"               "description"                    
 [5] "description_translated"          "funded_amount"                  
 [7] "loan_amount"                     "status"                         
 [9] "activity_name"                   "sector_name"                    
[11] "loan_use"                        "country_code"                   
[13] "country_name"                    "town_name"                      
[15] "currency_policy"                 "currency_exchange_coverage_rate"
[17] "currency"                        "partner_id"                     
[19] "posted_time"                     "planned_expiration_time"        
[21] "disburse_time"                   "raised_time"                    
[23] "lender_term"                     "num_lenders_total"              
[25] "num_journal_entries"             "num_bulk_entries"               
[27] "tags"                            "borrower_genders"               
[29] "borrower_pictured"               "repayment_interval"             
[31] "distribution_model"

In [6]:
dim(df1)

[1] 1419607      31

* El conjunto de datos consiste en 1419607 registros y 31 variables.



Según las utilizades para el registro de tiempos de ejecución de scripts en Jupyter Notebook (https://github.com/ipython-contrib/jupyter_contrib_nbextensions), tomó 23.6 segundos cargar sin error el conjunto de datos.

* **Consideración importante**: Aunque no toma mucho tiempo cargar el conjunto de datos ya que la función utiliza todos los núcleos (CORES) del computador, se puede notar que mantener el conjunto de datos abierto en R consume aproximadamente la mitad (3.37 GB de 7.65 GB)de la memoria RAM del computador (sección _Mem_ al costado inferior izquierdo de la imágen adjunta), limitando así la capacidad de cómputo al momento de procesar información. 

<img src="02.png">


### 3. Pruebas de eficiencia: (loops vs funciones vectorizadas)
Se puede apreciar que en la tabla que muestra los primeros 5 registros del conjunto de datos (_head(df1)_) se pueden apreciar dos columnas que llevan por nombre _description_ y _description__translated_. Si se desea realizar un estudio resumen del texto asociado a las descripciones de las personas que financian proyectos, el primer paso debería ser el obtener una columna única del texto en un solo idioma (inglés) para posteriormente realizar el trabajo de minería de datos textual.

Debido a que la columna _description__translated_ posee el texto traducido, el problema se aprecia en que si el texto original fue escrito en inglés, la nueva columna tiene valor nulo ya que todas las traducciones fueron adaptadas a inglés, por lo que una alternativa es verificar si es que la columna nueva tiene valor vacío y si es así traspasar el registro del idioma original y luego obtener la columna _description__translated_ completa.

#### 3.1 Estrategia ingenua (ciclos anidados)

La primera alternativa puede ser recorrer uno a uno cada registro de texto y revisar si es que existe registro nulo en _description__translated_, para luego llenarlo con el registro original (columna _description_).

In [7]:
for(i in 1:dim(df1)[1]){
    if(df1$description_translated[i] == ""){df1$description_translated[i] = df1$description[i]}
}

In [8]:
write.csv(df1,file = "data/df_backup.csv")

In [32]:
df1 = fread("data/df_backup.csv",header = T,sep = ",")
head(df1)

V1 loan_id loan_name      original_language
1 1  657307  Aivy           English          
2 2  657259  Idalia Marizza Spanish          
3 3  658010  Aasia          English          
4 4  659347  Gulmira        Russian          
5 5  656933  Ricky\\t       English          
6 6  659605  Gloria         English          
  description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
1 Aivy, 21 years of age, is single and lives in Ozamiz City, Misamis Occidental, Philippines. To help her parents meet their financial needs, she invests in a small sari-sari business (variety store). Her parents are thankful to her because of the things she has done for the family.\\r\\n\\r\\nIn the year 2013, Aivy became a member of PMPC. She believes that the institution can provide her financial assistance to access the loan amount of 5,000 PHP to buy additional inventories for her sari-sari store business. She is thankful to the cooperative in supporting her to grow in the small business she started.                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
2 Doña Idalia, esta casada, tiene 57 años de edad, vive con su esposo y sus 3 nietos en el municipio de Choloma, Cortes. Hace 7 años que se dedica a la venta de mercadería en general como ropa usada americana, lociones, calzado y otra achinería. Las ventas las realiza en forma ambulante. Ha solicitado 8,000 Lempiras para comprar ropa en general como pantalones, camisas, blusas, ropa de niño etc. para cumplir con los pedidos que tiene. Su propósito de seguir trabajando es poder ver en crecimiento su negocio.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
3 Aasia is a 45-year-old married lady and she has been running a small general store in her community for many years.  She offers a variety of grocery items at a reasonable price and has retained many customers who regularly shop at her store. Aasia makes enough to provide for her family.\\r\\n\\r\\nTo make her store fully stocked she requested a loan for 40,000 PKR which she will use to buy stock of rice, sugar and flour from the market at lower rates. In thi

In [33]:
tail(df1)

V1      loan_id loan_name          original_language
1 1419602 987606  Garik              English          
2 1419603 988180                                      
3 1419604 988213  Perlita            English          
4 1419605 989109  Okyeso Nyame Group English          
5 1419606 989143  Exequila           English          
6 1419607 989240  Lydia              French           
  description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
1 Garik, his parents and his sister live in Dalarik village in the Armavir region.<br /> <br />Garik works as an army officer and his father is an electrician. They work hard to sustain a living for the family. Simultaneously with his work, Garik also helps his family with farming work. His family's farm includes two cows, two oxen and hens. Near their house, Garik's family also owns arable land where apricots, watermelons, tomatoes, eggplants and peppers are grown. <br /><br />Garik has requested this loan to invest in the farm and prepare it for the winter season. He plans to purchase two oxen and fodder for feeding the livestock during the winter. <br /><br />This will be a big support for Garik and his family, and they all hope for lenders' support.                                                                                                                                                                                                                                                                                            
2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
3 Perlita is 52 years old, married and has three children.\\n\\nPerlita is in the agricultural business, pig fattening in the Philippines. Perlita requested a PHP 13000 loan through NWTF to buy feeds and other supplies to raise her pigs.\\n\\nPerlita has been in this business for 5 years now, and has borrowed 7 times in the past from NWTF to sustain it.\\n\\nIn the future, Perlita would like to save money to expand her business.                                         

In [34]:
length(which(df1$description_translated==""))/dim(df1)[1]

[1] 0.2949027

Luego de 6 hrs 21 minutos, el loop no ha sido capaz de recorrer el conjunto de datos completo y realizar los cambios. Solo pudo con el 29.5$%$ de la cantidad total de registros.

#### 3.2 Estrategia eficiente (funciones vectorizadas)

Dado que son demasiados registros como para recorrer uno a uno en el conjunto de datos completo, se puede considerar una función a la cual se le puedan ingresar vectores y condiciones que haga el mismo procedimiento de comprobar si es que existen registros nulos para luego obtener la columna final con texto. 

In [26]:
df1 = fread("data/loans.csv",header = T,sep = ",")
df1$description_translated = ifelse(df1$description_translated == "",df1$description,df1$description_translated)
head(df1)

loan_id loan_name      original_language
1 657307  Aivy           English          
2 657259  Idalia Marizza Spanish          
3 658010  Aasia          English          
4 659347  Gulmira        Russian          
5 656933  Ricky\\t       English          
6 659605  Gloria         English          
  description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
1 Aivy, 21 years of age, is single and lives in Ozamiz City, Misamis Occidental, Philippines. To help her parents meet their financial needs, she invests in a small sari-sari business (variety store). Her parents are thankful to her because of the things she has done for the family.\\r\\n\\r\\nIn the year 2013, Aivy became a member of PMPC. She believes that the institution can provide her financial assistance to access the loan amount of 5,000 PHP to buy additional inventories for her sari-sari store business. She is thankful to the cooperative in supporting her to grow in the small business she started.                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
2 Doña Idalia, esta casada, tiene 57 años de edad, vive con su esposo y sus 3 nietos en el municipio de Choloma, Cortes. Hace 7 años que se dedica a la venta de mercadería en general como ropa usada americana, lociones, calzado y otra achinería. Las ventas las realiza en forma ambulante. Ha solicitado 8,000 Lempiras para comprar ropa en general como pantalones, camisas, blusas, ropa de niño etc. para cumplir con los pedidos que tiene. Su propósito de seguir trabajando es poder ver en crecimiento su negocio.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
3 Aasia is a 45-year-old married lady and she has been running a small general store in her community for many years.  She offers a variety of grocery items at a reasonable price and has retained many customers who regularly shop at her store. Aasia makes enough to provide for her family.\\r\\n\\r\\nTo make her store fully stocked she requested a loan for 40,000 PKR which she will use to buy stock of rice, sugar and flour from the market at lower rates. In this way she would be ab

In [28]:
tail(df1)

loan_id loan_name          original_language
1 987606  Garik              English          
2 988180                                      
3 988213  Perlita            English          
4 989109  Okyeso Nyame Group English          
5 989143  Exequila           English          
6 989240  Lydia              French           
  description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
1 Garik, his parents and his sister live in Dalarik village in the Armavir region.<br /> <br />Garik works as an army officer and his father is an electrician. They work hard to sustain a living for the family. Simultaneously with his work, Garik also helps his family with farming work. His family's farm includes two cows, two oxen and hens. Near their house, Garik's family also owns arable land where apricots, watermelons, tomatoes, eggplants and peppers are grown. <br /><br />Garik has requested this loan to invest in the farm and prepare it for the winter season. He plans to purchase two oxen and fodder for feeding the livestock during the winter. <br /><br />This will be a big support for Garik and his family, and they all hope for lenders' support.                                                                                                                                                                                                                                                                                            
2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
3 Perlita is 52 years old, married and has three children.\\n\\nPerlita is in the agricultural business, pig fattening in the Philippines. Perlita requested a PHP 13000 loan through NWTF to buy feeds and other supplies to raise her pigs.\\n\\nPerlita has been in this business for 5 years now, and has borrowed 7 times in the past from NWTF to sustain it.\\n\\nIn the future, Perlita would like to save money to expand her business.                                                                                                 

In [29]:
length(which(df1$description_translated==""))

[1] 43285

Finalmente, existen 43285 registros sin texto.

### 4. Conclusiones:

* Si bien ambos procedimientos (3.1 y 3.2) son capaces de realizar la misma tarea, se ha comprobado una clara diferencia respecto a la eficiencia que involucra utilizar cada algoritmo.
* Utilizar loops, en contextos de una gran cantidad de registros, no es la manera más eficiente de realizar operaciones de búsqueda. Se puede ver en la siguiente figura que, además de requerir 6 hrs. y 21 minutos para ejecutar solo el 30$%$ inicial, el loop requiere la demanda de toda la memoria RAM (_Men_ en la esquina inferior izquierda) posible. Esto es una desventaja ya que no se puede disponer de memoria por un tiempo determinado y se corre el riesgo latente de que el computador se congele.

<img src="03.png">

* Efectivamente, utilizar funciones vectorizadas para recorrer vectores consume una menor cantidad de tiempo en comparación a loops y permite además realizar pruebas completas debido al poco tiempo que requiere la ejecución. En este caso, solo bastan 10.8 segundos para realizar todos los cambios.
* En este estudio no se han considerado loops anidados ya que el conjunto de datos tiene muchos registros por lo que es demasiada carga para el computador, pero se ha comprobado que las funciones vectorizadas son mucho más eficientes que un solo loop.